collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java

```diff
@@ -0,0 +1,248 @@
+
+public class DriveHelper {
+
+    private final Drive drive;
+
+    DriveHelper(@NonNull GoogleAccountCredential credential,
+                @NonNull HttpTransport transport,
+                @NonNull JsonFactory jsonFactory) {
+        drive = new Drive.Builder(transport, jsonFactory, credential)
+                .setApplicationName("ODK-Collect")
+                .build();
+    }
+
+    /**
+     * Constructs a new DriveHelper with the provided Drive Service.
+     * This Constructor should only be used for testing.
+     */
+    DriveHelper(@NonNull Drive drive) {
+        this.drive = drive;
+    }
+
+    /**
+     * Returns id of the root folder or null
+     */
+    public String getRootFolderId() throws IOException {
+        return drive.files()
+                .get("root")
+                .setFields("id")
+                .execute().getId();
+    }
+
+    public Drive.Files.List buildRequest(String query) {
+        try {
+            return drive.files()
+                    .list()
+                    .setQ(query)
+                    .setFields("nextPageToken, files(modifiedTime, id, name, mimeType)");
+
+        } catch (IOException e) {
+            Timber.e(e);
+        }
+        return null;
+    }
+
+    public void downloadFile(String fileId, FileOutputStream fileOutputStream) throws IOException {
+        try {
+            drive.files().get(fileId).executeMediaAndDownloadTo(fileOutputStream);
+        } finally {
+            try {
+                if (fileOutputStream != null) {
+                    fileOutputStream.close();
+                }
+            } catch (IOException e) {
+                Timber.e(e, "Unable to close the file output stream");
+            }
+        }
+    }
+
+    public String createOrGetIDOfFolderWithName(String jrFormId)
+            throws IOException, MultipleFoldersFoundException {
+        String submissionsFolderId = createOrGetIDOfSubmissionsFolder();
+        return getIDOfFolderWithName(jrFormId, submissionsFolderId, true);
+    }
+
+    private String createOrGetIDOfSubmissionsFolder()
+            throws IOException, MultipleFoldersFoundException {
+        String rootFolderId = getIDOfFolderWithName(GOOGLE_DRIVE_ROOT_FOLDER, null, true);
+        return getIDOfFolderWithName(GOOGLE_DRIVE_SUBFOLDER, rootFolderId, true);
+    }
+
+    /**
+     * Searches for the folder saved in google drive with the given name and returns it's id
+     *
+     * @param name                   The name of the folder saved in google drive
+     * @param inFolder               The id of the folder containing the given folder
+     * @param shouldCreateIfNotFound If the given folder is not found then this parameter decides
+     *                               whether to create a new folder with the given name or not
+     */
+    @Nullable
+    public String getIDOfFolderWithName(@NonNull String name, @Nullable String inFolder, boolean shouldCreateIfNotFound)
```

Comment:
You may want to consider having this class *only* make calls to Drive, e.g. these ones:

```
drive.files().get("root") .setFields("id") .execute().getId();
```

We can then have another class that contains business logic like this method—that way we can mock the class out that makes the Drive calls and test this logic here separately.

Not 100% necessary, but something to think about at least.